### PR TITLE
Implement support AS923 for RAKWireless 2245 PiHat.

### PR DESCRIPTION
**Pull request by sophiekovalevsky**
### Contribution description
Can be seen in https://github.com/brocaar/chirpstack-gateway-os/issues/41

This implementation will make possible to work with RAK2245 PiHat gateways at 923 MHz and includes the following tasks:

- [x] Create vendor rakwireless module, modify functions to read specific configuration model. Include model at region version at README. 
- [ ] ~~Packaging ipk. Include build in Makefile.~~ 
- [x] Update LUT table according to https://github.com/RAKWireless/rak_common_for_gateway/issues/12

### CODE REVIEW

File: chirpstack-concentratord-sx1301/src/config/vendor/mod.rs [hunk 1]
```diff
@@ -5,6 +5,7 @@ pub mod imst;
 pub mod kerlink;
 pub mod multitech;
 pub mod wifx;
+pub mod rakwireless;
```
Comment (brocaar) [inline chirpstack-concentratord-sx1301/src/config/vendor/mod.rs]: Maybe shorten this to just `rak` :)

File: chirpstack-concentratord-sx1301/src/config/vendor/rakwireless/mod.rs [hunk 2]
```diff
@@ -0,0 +1 @@
+pub mod rak2245_pi_hat_as923;
```
Comment (brocaar) [inline chirpstack-concentratord-sx1301/src/config/vendor/rakwireless/mod.rs]: Do you know if there is a difference between the rak2245 Pi Hat and other 2245 models? As far as I know it is exactly the same module, but soldered on a PCB that fits on a Pi. If there is no difference, then maybe there is no need to add the `_pi_hat` in the model name :)

**Comment by brocaar** [comment]
Is _Packaging ipk. Include build in Makefile._ needed? I started doing that for gateways that do not run the ChirpStack Gateway OS. For Raspberry Pi, I think there is no need to generate .ipk packages as this is already done with Yocto. The only thing would be to update the `gateway-config` script in the Gateway OS repo. That saves a bit of work :)

File: chirpstack-concentratord-sx1301/src/config/vendor/mod.rs (hunk 1 above)
Comment (sophiekovalevsky) [reply chirpstack-concentratord-sx1301/src/config/vendor/mod.rs]: Roger that :)

File: chirpstack-concentratord-sx1301/src/config/vendor/rakwireless/mod.rs (hunk 2 above)
Comment (sophiekovalevsky) [reply chirpstack-concentratord-sx1301/src/config/vendor/rakwireless/mod.rs]: I know that there are several hardware gateways that RAK offers with this model but I am not aware if all their products that use RAK2245 are the same in terms of pcb components, in any case, I would rather move taking away _pi_hat and if somebody uses another RAK2245 with some difference, then updated it as we move along.

Comment (brocaar) [reply chirpstack-concentratord-sx1301/src/config/vendor/rakwireless/mod.rs]: Sounds good to me :+1:

**Comment by sophiekovalevsky** [comment]
> Is _Packaging ipk. Include build in Makefile._ needed? I started doing that for gateways that do not run the ChirpStack Gateway OS. For Raspberry Pi, I think there is no need to generate .ipk packages as this is already done with Yocto. The only thing would be to update the `gateway-config` script in the Gateway OS repo. That saves a bit of work :)

That's why I initially wanted to make the pull request as draft :) since I was not sure about whether or not to implement this and I was requiring to check that in more detail. I am clear now about it. Thanks for letting me know.

**Comment by brocaar** [comment]
Yes, I did see that it was still a draft PR. Just wanted to make sure that you would not spend a lot of time on .ipk packaging.

**Comment by brocaar** [comment]
@sophiekovalevsky I believe RAK has made some updates: https://github.com/RAKWireless/rak_common_for_gateway/commit/099555865a42238f125c68ded5233a985747c40d :tada:

**Comment by sophiekovalevsky** [comment]
@brocaar I have tested again their new values and though I've seen a little bit of improvement, I still not getting results that state in rf power. I've seen some values that are so far away to what is expected and this really concerns me in a sense that the drift is just to much in my opinion. What are your thoughts?

**Comment by brocaar** [comment]
@sophiekovalevsky hmmm... that is not really assuring... Unfortunately I do not have the tools to test this, but I would recommend to share this with RAK. Ideally they should provide us with the correct calibration values, it should not be the other way around.

**Comment by sophiekovalevsky** [comment]
@brocaar understood. I will let RAK know about what I am experiencing with and see if I get some reply back. I think this feature can still in progress for at least the end of the week. Otherwise then I will update the LUT values with the latest values that they stated in their official repo.

**Comment by hdtodd** [comment]
I can't help on the technical issues around LUT values, but I did report an error in this thread:
    https://forum.rakwireless.com/t/error-packet-rejected-unsupported-rf-power-for-tx-17/1656
that I hope someone with expertise will look at.

There's no entry in the LUT tables for tx power 17, but it looks like TTN sends the gateway packets with that setting and lora_pkt_fwd rejects them based on power setting alone.  That's probably not the desired behavior.  

I was planning to set one of the LUTs to tx power 17 just to get packets through, but I don't know what problems that might cause.  If someone who know what that setting is supposed to do would examine the code and fix it or the LUT entries, I would appreciate it.  

It seems odd that the code filters based on tx power value ... there would be many settings in which downlink packet forwarding would fail.  Is "17" maybe supposed to be a wildcard setting?

**Comment by brocaar** [comment]
@sophiekovalevsky how would you like to move forward with this pull-request? I'm not sure when RAK will respond to your issue. For the RAK2246 I'm experiencing the same issue and I have created a ticket for that too (https://github.com/RAKWireless/rak_common_for_gateway/issues/13).

If you agree, then I'm happy to merge this so we can at least move forward with the ChirpStack Gateway OS pull-request :)

**Comment by hdtodd** [comment]
My apologies to you both: this problem was fixed a couple of weeks ago.  

Though I had cloned and installed the latest master from github, I apparently was running an old version when I found that tx power 17 downlink packets weren't being forwarded.

In trying to debug my inability to get lora_pkt_fwd to read the GPS, I had reason to go through it all again.  And I now see that the code changes tx power 17 requests to 16 and processes them.  I've run the current version and confirmed that this is no longer a problem.

I'm sure you guys are busy enough doing real work that you don't need distractions like this one.  My apologies.

David

**Comment by sophiekovalevsky** [comment]
@brocaar,

I've received a reply back from RAK and pushed LUT table changes provided in their dev branch. I saw that have created https://github.com/brocaar/chirpstack-concentratord/issues/6 where configuration points to master. Does this change conflicts with it? I think that there are soon push to changes into mainstream.

**Comment by brocaar** [comment]
@sophiekovalevsky you're right #6 should have pointed to the dev branch :)
